packages/core/src/modules/connections/DidExchangeProtocol.ts

```
@@ -0,0 +1,462 @@
+        recipientKeys: didDocument.recipientKeys,
+      },
+    })
+
+    await this.didRepository.save(didRecord)
+
+    connectionRecord.theirDid = message.did
+    connectionRecord.theirLabel = message.label
+    connectionRecord.threadId = message.id
+
+    await this.updateState(DidExchangeRequestMessage.type, connectionRecord)
+    this.logger.debug(`Process message ${DidExchangeRequestMessage.type} end`, connectionRecord)
+    return connectionRecord
+  }
+
+  public async createResponse(
+    connectionRecord: ConnectionRecord,
+    routing?: Routing
+  ): Promise<DidExchangeResponseMessage> {
+    this.logger.debug(`Create message ${DidExchangeResponseMessage.type} start`, connectionRecord)
+    DidExchangeStateMachine.assertCreateMessageState(DidExchangeResponseMessage.type, connectionRecord)
+
+    // They will then either update the provisional service information to rotate the key, or provision a new DID entirely.
+    // The choice here will depend on the nature of the DID used in the invitation.
+
+    // if reuse did from invitation then do ...
+    // otherwise create new did and didDoc
+    const { did, threadId } = connectionRecord
+
+    if (!threadId) {
+      throw new AriesFrameworkError('Missing threadId on connection record.')
+    }
+
+    // The responder will then craft an exchange response using the newly updated or provisioned information.
+
+    // Sign message attachment
+    // Use invitationKey by default, fall back to verkey (?)
+    const [verkey] = connectionRecord.invitation?.recipientKeys || []
+
+    if (!verkey) {
+      throw new AriesFrameworkError('Connection invitation does not contain recipient key.')
+    }
+
+    const peerDidRouting = routing || {
+      endpoints: connectionRecord.invitation?.serviceEndpoint ? [connectionRecord.invitation?.serviceEndpoint] : [],
+      verkey,
+      did,
+      routingKeys: connectionRecord.invitation?.routingKeys || [],
```

Same here, if we don't have an invitation we should throw an error